### PR TITLE
Partial refactor

### PR DESCRIPTION
**Pull request by ajlee21**
This PR refactors scripts to attempt to modularize functions:

In this PR the following was done:
1.  Scripts pertaining to earlier exploratory analyses were moved into a new directory called archive/
2.  A main script was created to perform the PA1673 gradient analysis (an analysis to predict the gene expression profile along a continuum using manipulations in latent space and gene space)
3.  Created functions from previous notebooks to streamline analysis
4.  Created a separate pre-processing function
5.  Added output files from PA1673 gradient analysis

### CODE REVIEW

File: scripts/functions/def_offset.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
+#------------------------------------------------------------------------------------------------------------------------------
+# By Alexandra Lee 
+# (updated October 2018) 
+#
+# Define offset vectors
```
Comment (gwaybio) [inline scripts/functions/def_offset.py:5]: this docstring needs more description. What is an offset vector? How do I use the script? Is there an expected output?

File: scripts/functions/def_offset.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+#-------------------------------------------------------------------------------------------------------------------------------
+import os
+import pandas as pd
+import numpy as np
+from keras.models import model_from_json, load_model
+from keras import metrics, optimizers
+
+randomState = 123
+from numpy.random import seed
```
Comment (gwaybio) [inline scripts/functions/def_offset.py:16]: all import statements should come before any other execution, if possible. (randomState should be defined after importing `seed`)

File: scripts/functions/def_offset.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+seed(randomState)
+
+def gene_space_offset(data_dir, gene_id):
+    """
+    Gene space offset:
+
+    offset_vector = average(gene expression of samples that have the highest 5% of target gene expression) -
```
Comment (gwaybio) [inline scripts/functions/def_offset.py]: what is "target gene expression"?

File: scripts/functions/def_offset.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    
+    # Load arguments
+    target_gene_file = os.path.join(data_dir, "PA1673.txt")
+    non_target_gene_file = os.path.join(data_dir, "train_model_input.txt.xz")
+
+    # Output files
+    offset_file = os.path.join(data_dir, "offset_gene_space.txt")
+    lowest_file = os.path.join(data_dir, "lowest.txt")
+    highest_file = os.path.join(data_dir, "highest.txt")
+
```
Comment (gwaybio) [inline scripts/functions/def_offset.py:58]: extra space

Comment (gwaybio) [reply scripts/functions/def_offset.py:58]: lines 44 and 48 too

File: scripts/functions/def_offset.py (hunk 3 above)
Comment (gwaybio) [reply scripts/functions/def_offset.py]: also, what is the format of `gene_id`?

File: scripts/functions/def_offset.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    target_gene_data = pd.read_table(target_gene_file, header=0, sep='\t', index_col=0)
+    non_target_gene_data = pd.read_table(non_target_gene_file, header=0, sep='\t', index_col=0)
+
+    # read in saved models
+    loaded_model = load_model(model_file)
+
+    # load weights into new model
+    loaded_model.load_weights(weights_file)
+
+    # Sort target gene data by expression (lowest --> highest)
```
Comment (gwaybio) [inline scripts/functions/def_offset.py:145]: it looks like a lot of this code is duplicated from the previous function. It might be worth splitting out the thresholding bits into its own function

File: scripts/functions/generate_input.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+    # Output
+    train_input_file = os.path.join(base_dir, "train_model_input.txt.xz")
+    PA1673_file = os.path.join(base_dir, "PA1673.txt")
+
+
+    # Read in data
+    data = pd.read_table(data_file, header=0, sep='\t', index_col=0, compression='zip')
+    X = data.transpose()
+    X.head(5)
```
Comment (gwaybio) [inline scripts/functions/generate_input.py]: probably don't want to call this inside a function

Comment (gwaybio) [reply scripts/functions/generate_input.py]: there are a couple of other instances of similar things in this script too. Are they debugging relics?

File: scripts/functions/interpolate.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+    # Read in data
+    target_gene_data = pd.read_table(target_gene_file, header=0, sep='\t', index_col=0)
+    non_target_gene_data = pd.read_table(non_target_gene_file, header=0, sep='\t', index_col=0)
+    offset_data = pd.read_table(offset_file, header=0, sep='\t', index_col=0)
+
+    # Sort target gene data by expression (lowest --> highest)
+    target_gene_sorted = target_gene_data.sort_values(by=[gene_id])
+
+    # Get sample IDs with the lowest 5% of target gene expression
```
Comment (gwaybio) [inline scripts/functions/interpolate.py:74]: separating out a thresholding function will help here too

File: scripts/functions/def_offset.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+
+    offset_latent_space_df = pd.Series.to_frame(offset_latent_space).T
+
+    # output lowest and highest expressing samples
+    low_exp_encoded_df.to_csv(lowest_file, sep='\t')
+    high_exp_encoded_df.to_csv(highest_file, sep='\t')
+
+    # ouput gene space offset vector
+    offset_latent_space_df.to_csv(offset_file, sep='\t')
```
Comment (dhimmel) [inline scripts/functions/def_offset.py]: newline at end of file

File: scripts/functions/generate_input.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    data_file = os.path.join(os.path.dirname(os.getcwd()), "data", "all-pseudomonas-gene-normalized.zip")
+    gene_id = "PA1673"
+
+    # Output
+    train_input_file = os.path.join(base_dir, "train_model_input.txt.xz")
+    PA1673_file = os.path.join(base_dir, "PA1673.txt")
+
+
+    # Read in data
+    data = pd.read_table(data_file, header=0, sep='\t', index_col=0, compression='zip')
```
Comment (dhimmel) [inline scripts/functions/generate_input.py]: read_table defaults to tab sep.

File: scripts/functions/interpolate.py [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+        true = pd.Series.to_frame(non_target_gene_data.loc[sample_id]).T
+
+        [coeff, pval] = pearsonr(predict.values.T, true.values.T)
+        corr_score[sample_id] = coeff 
+
+    corr_score_df = pd.DataFrame.from_dict(corr_score, orient='index')
+
+    # Output estimated gene experession values
+    corr_score_df.to_csv(corr_file, sep='\t')
```
Comment (dhimmel) [inline scripts/functions/interpolate.py]: missing newline

File: stats/PA1673_gradient/tybalt_2layer_10latent_stats.csv [hunk 11]
```diff
@@ -0,0 +1,201 @@
+	loss	val_loss	learning_rate	batch_size	epochs	kappa
```
Comment (dhimmel) [inline stats/PA1673_gradient/tybalt_2layer_10latent_stats.csv:1]: This should be a .tsv not .csv. Also seems like the index column should not be saved here

**Review by dhimmel — COMMENTED** [review]
`*.pyc` files should be ignored.

`.tsv` extension is preferred for tab-delimited text over `.txt`

why is this pr titled "continuous analysis"... I don't see any continuous integration

`viz/PA1673_gradient/corr.png` is nice!

File: scripts/functions/interpolate.py [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+
+    corr_score = {}
+    for sample_id in remaining_ids:
+        intermediate_target_gene_exp = target_gene_sorted.loc[sample_id]
+        alpha = utils.get_scale_factor(target_gene_file, gene_id, intermediate_target_gene_exp)
+
+        predict = baseline + alpha.values[0]*offset_data
+        true = pd.Series.to_frame(non_target_gene_data.loc[sample_id]).T
+
+        [coeff, pval] = pearsonr(predict.values.T, true.values.T)
```
Comment (gwaybio) [inline scripts/functions/interpolate.py:96]: Is it worth saving pvals here too? Can add to the `corr_score` dict

File: scripts/functions/interpolate.py [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+
+    corr_score_df = pd.DataFrame.from_dict(corr_score, orient='index')
+
+    # Output estimated gene experession values
+    corr_score_df.to_csv(corr_file, sep='\t')
+    target_gene_sorted.to_csv(sorted_file, sep='\t')
+
+def interpolate_latent_space(data_dir, model_dir, encoded_dir, gene_id, out_dir):
+    """
+    Interpolate in gene space:
```
Comment (gwaybio) [inline scripts/functions/interpolate.py]: "in latent space"

File: scripts/functions/interpolate.py [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+    # Use trained model to encode expression data into SAME latent space
+    low_exp_encoded = loaded_model.predict_on_batch(low_exp)
+    low_exp_encoded_df = pd.DataFrame(low_exp_encoded, index=low_exp.index)
+
+    # Average gene expression across samples in each extreme group
+    lowest_mean_encoded = low_exp_encoded.mean(axis=0)
+
+    # Format and rename as "baseline"
+    baseline_encoded = pd.DataFrame(lowest_mean_encoded, index=offset_encoded.columns).T
+    baseline_encoded
```
Comment (gwaybio) [inline scripts/functions/interpolate.py]: remove

File: scripts/functions/utils.py [hunk 15]
```diff
@@ -0,0 +1,47 @@
+#----------------------------------------------------------------------------------------------------------------
+# Helper functions
+#----------------------------------------------------------------------------------------------------------------
+import pandas as pd
+import numpy as np
+import os
+
+def get_scale_factor (ref_file, gene_id, expression_profile):
+
```
Comment (gwaybio) [inline scripts/functions/utils.py:9]: extra lines

File: scripts/functions/utils.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+    """
+    Determine how much to scale offset based on distance along the target gene expression gradient
+    
+    Output:
+     scale factor = intermediate gene expression/ (average high target gene expression - avgerage low target gene expression) 
+    """
+
+    # Read in reference gene expression file
+    ref_data = pd.read_table(ref_file, header=0, sep='\t', index_col=0)
```
Comment (gwaybio) [inline scripts/functions/utils.py]: i noticed this is being called in all the loops. Instead of reading in the data each time, why not include `ref_data` as a function argument and read in before looping

File: scripts/functions/vae.py [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+K.set_session(sess)
+
+
+from keras.layers import Input, Dense, Lambda, Layer, Activation
+from keras.layers.normalization import BatchNormalization
+from keras.models import Model, Sequential
+from keras import metrics, optimizers
+from keras.callbacks import Callback
+
+def tybalt_2layer_model(learning_rate, batch_size, epochs, kappa, intermediate_dim, latent_dim, epsilon_std, base_dir, analysis_name):
```
Comment (gwaybio) [inline scripts/functions/vae.py:64]: may be worth checking out [`tybalt.data_models`](https://github.com/greenelab/tybalt/blob/master/tybalt/data_models.py#L179).

Can use like `from tybalt import DataModel`. I don't have a 2 layer mode here, but contributions are welcome!

File: scripts/functions/generate_input.py (hunk 9 above)
Comment (ajlee21) [reply scripts/functions/generate_input.py]: fixed

File: scripts/functions/interpolate.py (hunk 10 above)
Comment (ajlee21) [reply scripts/functions/interpolate.py]: fixed

File: stats/PA1673_gradient/tybalt_2layer_10latent_stats.csv (hunk 11 above)
Comment (ajlee21) [reply stats/PA1673_gradient/tybalt_2layer_10latent_stats.csv:1]: fixed

File: scripts/functions/def_offset.py (hunk 2 above)
Comment (ajlee21) [reply scripts/functions/def_offset.py:16]: fixed

File: scripts/functions/def_offset.py (hunk 1 above)
Comment (ajlee21) [reply scripts/functions/def_offset.py:5]: very good point, docstring is updated

File: scripts/functions/def_offset.py (hunk 3 above)
Comment (ajlee21) [reply scripts/functions/def_offset.py]: This information is now added.  The target gene is used to sort the samples and can be thought of as your "phenotype".  In "interpolate_<>.py", after we sort samples based on the expression level of the target gene, we want to predict the expression profile of the OTHER genes at different levels of target gene expression.  Hopefully that is a bit more clear.  It is a bit confusing because we are using expression to predict expression.

File: scripts/functions/def_offset.py (hunk 4 above)
Comment (ajlee21) [reply scripts/functions/def_offset.py:58]: fixed

File: scripts/functions/generate_input.py (hunk 6 above)
Comment (ajlee21) [reply scripts/functions/generate_input.py]: yes, good eye

File: scripts/functions/interpolate.py (hunk 12 above)
Comment (ajlee21) [reply scripts/functions/interpolate.py:96]: I am not doing anything with the p-values at the moment but I can see use for this in the future if we are able to get good differentiation between gene space vs latent space.  In that case I this is a good suggestion

File: scripts/functions/interpolate.py (hunk 14 above)
Comment (ajlee21) [reply scripts/functions/interpolate.py]: fixed

File: scripts/functions/utils.py (hunk 15 above)
Comment (ajlee21) [reply scripts/functions/utils.py:9]: fixed

File: scripts/functions/interpolate.py (hunk 13 above)
Comment (ajlee21) [reply scripts/functions/interpolate.py]: changed

File: scripts/functions/interpolate.py (hunk 7 above)
Comment (ajlee21) [reply scripts/functions/interpolate.py:74]: Added a new function for this computation

File: scripts/functions/vae.py (hunk 17 above)
Comment (ajlee21) [reply scripts/functions/vae.py:64]: This is definitely a good idea and something started to think about during this refactor, however I think this change will need to wait in order to get additional analysis completed.  Hopefully in the next refactor.

File: scripts/functions/utils.py (hunk 16 above)
Comment (ajlee21) [reply scripts/functions/utils.py]: This is a good point, I have changed this to read in the objected that is already read in instead of read it in each time through the for loop

File: encoded/PA1673_gradient/lowest_encoded.txt [hunk 18]
```diff
@@ -0,0 +1,61 @@
+	0	1	2	3	4	5	6	7	8	9
+Glu_6Hour_(Pae_G1a).CEL	6.7398786544799805	0.0	0.0	5.70363187789917	0.0	6.582232475280762	0.0	0.0	0.0	0.0
+JS-A84.9.07.CEL	8.86330795288086	0.0	0.0	0.9941893219947815	0.13427108526229858	4.0595221519470215	0.0	0.0	0.0	0.0
```
Comment (dhimmel) [inline encoded/PA1673_gradient/lowest_encoded.txt:3]: If you're saving these files with pandas, I'd do `float_format="%.5g"` to cut down on the excessive decimal places.
